packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+  noChange,
+  Directive,
+  Part,
+  NODE_PART,
+  AttributePart,
+} from 'lit-html';
+
+import {$private} from 'lit-html/private-ssr-support.js';
+
+const {getTemplateHtml, marker, markerMatch, boundAttributeSuffix} = $private;
+
+import {digestForTemplateResult} from 'lit-html/hydrate.js';
+
+import {ElementRenderer} from './element-renderer.js';
+
+import {escapeAttribute, escapeTextContent} from './util/escaping.js';
+
+// types only
+import {DefaultTreeDocumentFragment} from 'parse5';
+
+import {
+  traverse,
+  parseFragment,
+  isCommentNode,
+  isElement,
+} from './util/parse5-utils.js';
+
+import {isRenderLightDirective} from 'lit-html/directives/render-light.js';
+import {LitElement} from 'lit-element';
+import {LitElementRenderer} from './lit-element-renderer.js';
+import {reflectedAttributeName} from './reflected-attributes.js';
+
+declare module 'parse5' {
+  interface DefaultTreeElement {
+    isDefinedCustomElement?: boolean;
+  }
+}
+
+Directive.prototype.resolve = function (
+  this: Directive,
+  _part: Part,
+  values: unknown[]
+) {
+  return this.render(...values);
+};
+
+const templateCache = new Map<
+  TemplateStringsArray,
+  {
+    ops: Array<Op>;
+    attrNames: Array<string>;
+  }
+>();
+
+/**
+ * Operation to output static text
+ */
+type TextOp = {
+  type: 'text';
+  value: string;
+};
+
+/**
+ * Operation to output dynamic text from the associated template result value
+ */
+type NodePartOp = {
+  type: 'node-part';
+  index: number;
+  useCustomElementInstance?: boolean;
+};
+
+/**
+ * Operation to output an attribute with bindings. Includes all bindings for an
+ * attribute, like an attribute template part or AttributeComitter.
+ */
+type AttributePartOp = {
+  type: 'attribute-part';
+  index: number;
+  name: string;
+  strings: Array<string>;
+  tagName: string;
+  useCustomElementInstance?: boolean;
+};
+
+/**
+ * Operator to create a custom element instance.
+ */
+type CustomElementOpenOp = {
+  type: 'custom-element-open';
+  tagName: string;
+  ctor: {new (): HTMLElement};
+  staticAttributes: Map<string, string>;
+};
+
+/**
+ * Operation to render a custom element's attributes. This is separate from
+ * `custom-element-open` because attribute/property parts go in between and need
+ * to run and be set on the instance before we render the element's final
+ * attributes.
+ */
+type CustomElementAttributesOp = {
+  type: 'custom-element-attributes';
+};
+
+/**
+ * Operation to render a custom element's children, usually its shadow root.
+ */
+type CustomElementChildrenOp = {
+  type: 'custom-element-children';
+};
+
+/**
+ * Operation to close a custom element so that its no longer available for
+ * bindings.
+ */
+type CustomElementClosedOp = {
+  type: 'custom-element-close';
+};
+
+type Op =
+  | TextOp
+  | NodePartOp
+  | AttributePartOp
+  | CustomElementOpenOp
+  | CustomElementAttributesOp
+  | CustomElementChildrenOp
+  | CustomElementClosedOp;
+
+const getTemplate = (result: TemplateResult) => {
```

Comment:
Add comment about what this is doing, including that it's creating op codes.

Reply:
See note above about how it's currently the element renderer implementation's responsibility to render the `<template shadowroot>`, so right now this naming is correct imo. We could limit the only thing an element renderer is allowed to render is shadow root children, in which case we can make rendering the `<template shadowroot>` part of the core.

Reply:
Worth a comment, but since this is porting over the existing code I think we can do some passes afterwards. The key part here is that this is essentially the SSR version of lit-html's `getTemplate()`/`new Template()`. We could structure it to match more closely as well.

Reply:
(whoops, comment got was out of place somehow)

Added comments.